app/controllers/visualizations_controller.rb

```diff
@@ -214,7 +214,7 @@ def background_for_heatmap
 
   # The most recent update time of any samples pipeline run.
   def heatmap_ts
-    updated_ats = samples_for_heatmap.map { |sample| sample.first_pipeline_run.updated_at }
+    updated_ats = samples_for_heatmap.map { |sample| sample.first_pipeline_run && sample.first_pipeline_run.updated_at }
```

Comment:
You are running the query twice here (inside first_pipeline_run)
Load the pipeline run and then use it...

Reply:
I was wondering about that... does activerecord always re-execute a query?

Reply:
From the logs and the docs, it looks like the second call **is** cached: https://edgeguides.rubyonrails.org/caching_with_rails.html#sql-caching . Are we trying to avoid relying on the cache?

Reply:
In this case it might execute only once, since the `pipeline_runs` are being included in `samples_for_heatmap` query (see my comment below)